pkg/client/block/client.go

```diff
@@ -0,0 +1,169 @@
+package block
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"cosmossdk.io/depinject"
+
+	"pocket/pkg/client"
+	"pocket/pkg/either"
+	"pocket/pkg/observable"
+	"pocket/pkg/observable/channel"
+)
+
+var (
+	_ client.BlockClient = (*blockClient)(nil)
+	_ client.Block       = (*cometBlockEvent)(nil)
+)
+
+// blockClient implements the BlockClient interface.
+type blockClient struct {
+	// endpointURL is the URL of RPC endpoint which eventsClient subscription
+	// requests will be sent.
+	endpointURL string
+	// eventsClient is the events query client which is used to subscribe to
+	// newly committed block events. It emits an either value which may contain
+	// an error, at most, once and closes immediately after if it does.
+	eventsClient client.EventsQueryClient
+	// latestBlockObsvblsReplay is a replay observable with replay buffer size 1,
+	// which holds the "active latest block observable" which is notified when
+	// block commit events are received by the events query client subscription
+	// created in goPublishBlocks. This observable (and the one it emits) closes
+	// when the events bytes observable returns an error and is updated with a
+	// new "active" observable after a new events query subscription is created.
+	latestBlockObsvblsReplay observable.ReplayObservable[client.BlocksObservable]
+	// latestBlockObsvblsReplayPublishCh is the publish channel for latestBlockObsvblsReplay.
+	// It's used to set blockObsvbl initially and subsequently update it, for
+	// example, when the connection is re-established after erroring.
+	latestBlockObsvblsReplayPublishCh chan<- client.BlocksObservable
+}
+
+func NewBlockClient(
+	ctx context.Context,
+	deps depinject.Config,
+	cometWebsocketURL string,
+) (client.BlockClient, error) {
+	// Initialize block client
+	bClient := &blockClient{endpointURL: cometWebsocketURL}
+	bClient.latestBlockObsvblsReplay, bClient.latestBlockObsvblsReplayPublishCh =
+		channel.NewReplayObservable[client.BlocksObservable](ctx, 1)
+
+	// Inject dependencies
+	if err := depinject.Inject(deps, &bClient.eventsClient); err != nil {
```

Comment:
Wow. So all we do is just inject an interface into a `depinject.Config` and we can supply it elsewhere?

Reply:
Yessir. :smile:

Reply:
🤯

Reply:
In fact, on the provide/supply side it doesn't even have to be an object with an interface type. You can provide a concrete type and it will try to figure it out. There is also a mechanism to assist it with this mapping if it struggles.